bionic/core/task_execution.py

```diff
@@ -782,97 +785,126 @@ class RemoteSubgraph:
+                else:
+                    self.target_states.add(state)
+
+            if state.func_attrs.aip_task_config is not None:
+                self.states_with_aip_task_configs.add(state)
+            if not state.task.can_be_serialized:
+                self.non_serializable_states.add(state)
+
+            for dep_state in state.dep_states:
+                add_state_and_traverse(dep_state, is_dependency=True)
+            for followup_state in state.followup_states:
+                add_state_and_traverse(followup_state)
+
+        add_state_and_traverse(target_state)
+
+    def strip_states(self, states):
+        """
+        Returns copies of the provided TaskStates with any unnecessary state and
+        ancestors "stripped" off; these copies can be safely transmitted to another
+        process for computation.
+        """
+
+        stripped_states_by_task_key = {}
 
-        # Make a copy of the TaskState, which we'll strip down to make it easier to
-        # serialize.
-        # (This is a shallow copy, so we'll make sure to avoid mutating any of its
-        # member variables.)
-        stripped_state = copy.copy(original_state)
-        self._stripped_states_by_task_key[task_key] = stripped_state
+        def strip_state(original_state):
+            """Returns a stripped copy of a TaskState."""
 
-        # Strip out data cached in memory -- we can't necessarily pickle it, so
-        # we need to get rid of it before trying to transmit this state to another
-        # process.
-        stripped_state._result = None
+            task_key = original_state.task_key
+            if task_key in stripped_states_by_task_key:
+                return stripped_states_by_task_key[task_key]
 
-        # These fields are picklable, but only needed for cache setup and
-        # initialization.
-        if stripped_state.is_initialized:
-            stripped_state.case_key = None
+            assert original_state in self.all_states
+            assert original_state not in self.non_serializable_states
 
-        if stripped_state.should_persist:
-            assert len(stripped_state.followup_states) == 0
+            # Make a copy of the TaskState, which we'll strip down to make it
+            # easier to serialize.
+            # (This is a shallow copy, so we'll make sure to avoid mutating any of
+            # its member variables.)
+            stripped_state = copy.copy(original_state)
+            stripped_states_by_task_key[task_key] = stripped_state
 
-            if stripped_state.is_cached_persistently:
+            # Strip out data cached in memory -- we can't necessarily pickle it, so
+            # we need to get rid of it before trying to transmit this state to
+            # another process.
+            stripped_state._result = None
+
+            # These fields are picklable, but only needed for cache setup and
+            # initialization.
+            if stripped_state.is_initialized:
```

Comment:
Wondering, what scenario would the state be not serialized?

Reply:
I'm assuming you meant "in what scenario would the state not be initialized?" That's a great question and I had to poke around for a while to answer it. At first I thought this was impossible and this `if` statement was obsolete, because of this comment in `flow_execution`:

```
# We want every task state to be initialized before we try running it
# remotely. We'll just require this of the target states, since all the
# other ones are ancestors of them.
```

However, it turns out this comment is wrong (which I found out by adding `assert stripped_state.is_initialized` to the line you're asking about). In fact, we can have non-persistable followup tasks that are not initialized yet. For example:

```
@builder
@bn.persist(False)
@bn.outputs("a", "b")
def _():
    return 1, 2

@builder
def c(a):
    return a + 1

builder.build().get("c")
```

In this case, `b` is a follow-up of `a, b`, so it will appear in the subgraph; but it is not persistable, so it won't appear in `target_states` and won't get initialized.

I will fix that comment -- thanks!

(Also, as @namanjain pointed out, we probably don't need the `case_key` attribute at all more, so this line of code is obsolete for other reasons. I'm also not 100% sure we still need `every task state to be initialized before we try running it remotely`, but it doesn't hurt to do it, so I'll just add a `TODO` to come back and check if it's still necessary.)